### PR TITLE
Bael 5961: OpenTelemetry integration with Spring Boot application

### PR DESCRIPTION
**Pull request by saikatcse03**

### CODE REVIEW

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM adoptopenjdk/openjdk11:alpine
+
+COPY target/spring-cloud-open-telemetry1-1.0.0-SNAPSHOT.jar spring-cloud-open-telemetry1-1.0.0-SNAPSHOT.jar
```
Comment (andrzejsydor) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/Dockerfile:3]:
```suggestion
COPY target/spring-cloud-open-telemetry1-1.0.0-SNAPSHOT.jar spring-cloud-open-telemetry.jar
```

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/api/client/PriceClient.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.client.RestTemplate;
+
+@Component
+public class PriceClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(PriceClient.class);
+
+    @Autowired
+    private RestTemplate restTemplate;
```
Comment (andrzejsydor) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/api/client/PriceClient.java:19]: could be `final` and injected by constructor

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/controller/ProductController.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class ProductController {
+    private final Logger logger = LoggerFactory.getLogger(ProductController.class);
+
+    @Autowired
+    private PriceClient priceClient;
```
Comment (andrzejsydor) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/controller/ProductController.java:18]: also could be `final`

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/repository/ProductRepository.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+        if(!productMap.containsKey(productId)){
+            logger.error("Product Not Found for Product Id {}", productId);
+            throw new ProductNotFoundException("Product Not Found");
+        }
+
+        return productMap.get(productId);
+    }
+
+    private void setupRepo() {
```
Comment (andrzejsydor) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/repository/ProductRepository.java:34]: could be better @PostConstruct

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        mockMvc.perform(get("/price/" + productId))
+          .andExpect(status().is(HttpStatus.OK.value()));
+    }
+
+
+    @Test
+    public void givenProductNotFound_whenGetProductCalled_thenReturnInternalServerError() throws Exception {
+        long productId = 100000L;
+        Price price = new Price();
```
Comment (andrzejsydor) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java:52]: same as the price in the line 37, could be extracted to the method

**Review by andrzejsydor — CHANGES_REQUESTED** [review]
.

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/repository/ProductRepository.java (hunk 4 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/repository/ProductRepository.java:36]: Changed to postContrucut

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/Dockerfile (hunk 1 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/Dockerfile]: Removed version number

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/api/client/PriceClient.java (hunk 2 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/api/client/PriceClient.java]: done as mentioned in all places

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/controller/ProductController.java (hunk 3 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/controller/ProductController.java]: done

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java (hunk 5 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java]: Extracted the method in this class, as well in all similar common code

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/SpringContextTest.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+
+@ExtendWith(SpringExtension.class)
+@SpringBootTest(classes = ProductApplication.class)
+public class SpringContextTest {
+    
+    @Test
+    public void whenSpringContextIsBootstrapped_thenNoExceptions() {
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/SpringContextTest.java:14]: JUnit 5 test functions and classes don't use the `public` modifier - we can remove this and leave them at default visibility.

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+
+@ExtendWith(SpringExtension.class)
+@WebMvcTest(ProductController.class)
+public class ProductControllerUnitTest {
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java:27]: Let's remove all the `public`s here too.

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        Product product = getProduct(productId);
+
+        when(productRepository.getProduct(productId)).thenReturn(product);
+        when(priceCLient.getPrice(productId)).thenThrow(HttpServerErrorException.ServiceUnavailable.class);
+
+        mockMvc.perform(get("/product/" + productId))
+                .andExpect(status().is(HttpStatus.INTERNAL_SERVER_ERROR.value()));
+    }
+
+    private static Product getProduct(long productId) {
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java:78]: arguably, this is a `create` not a `get` method.

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/SpringContextTest.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+@ExtendWith(SpringExtension.class)
+@SpringBootTest(classes = PriceApplication.class)
+public class SpringContextTest {
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/SpringContextTest.java:11]: Remove `public` from this class

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/controller/PriceControllerUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.test.web.servlet.MockMvc;
+
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+
+@ExtendWith(SpringExtension.class)
+@WebMvcTest(PriceController.class)
+public class PriceControllerUnitTest {
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/controller/PriceControllerUnitTest.java:24]: and remove `public` here too

**Review by ashleyfrieze — CHANGES_REQUESTED** [review]

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/api/client/PriceClient.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.client.RestTemplate;
+
+@Component
+public class PriceClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(PriceClient.class);
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/api/client/PriceClient.java:16]: `logger` should be `LOGGER` if it's `private static final` (and I think it should be).

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java (hunk 7 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/controller/ProductControllerUnitTest.java]: Done here.

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/SpringContextTest.java (hunk 6 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/test/java/com/baeldung/opentelemetry/SpringContextTest.java]: removed public

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/SpringContextTest.java (hunk 9 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/SpringContextTest.java]: removed

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/controller/PriceControllerUnitTest.java (hunk 10 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry2/src/test/java/com/baeldung/opentelemetry/controller/PriceControllerUnitTest.java]: yes done

File: spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/api/client/PriceClient.java (hunk 11 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-open-telemetry/spring-cloud-open-telemetry1/src/main/java/com/baeldung/opentelemetry/api/client/PriceClient.java]: I've made it change though not sure its convention.
Generally i've used as Logger as its not constant data rather just a final static object reference. https://www.slf4j.org/api/org/slf4j/Logger.html.

**Pull request merged**